feat: add map service

JihwanB opened the pull request:
## ✨ 변경 타입
* [x] 신규 기능 추가/수정
* [ ] 버그 수정
* [ ] 리팩토링
* [ ] 설정
* [ ] 비기능 (주석 등 기능에 영향을 주지 않음)

## ✅ 체크리스트

* [ ] 코드 작성 가이드라인을 준수했는가?
* [ ] 변경 사항에 대한 테스트를 완료했는가?
* [ ] 문서 변경이 필요한 경우, 해당 내용을 반영했는가?

## 🚀 변경 내용
* **as-is**
  * N/A

* **to-be**
  * 각 서비스에서 호출하여 지도 관련 정보를 받을 수 있습니다.

## 💡 추가 설명

* 주소를 요청으로 받아 좌표를 반환해주는 geocode API와 좌표를 기반으로 출발지점과 도착지점의 경로 정보를 반환해주는 route API를 추가했습니다.
* 공통 코드는 건너뛰고 봐주셔도 무방할 듯 합니다.
* 수정사항 알려주시면 적극 반영하겠습니다!

## ⚡️ 관련 이슈

* #6 

## 📚 참고 자료 (선택 사항)

* https://api.ncloud-docs.com/docs/ai-naver-mapsgeocoding-geocode
* https://api.ncloud-docs.com/docs/ai-naver-mapsdirections-driving

JihwanB committed:
ci: add map-service to docker-compose.yml

- docker-compose.yml에 map-service 추가
- map-service Dockerfile 추가

JihwanB committed:
feat: initialize map-service

JihwanB committed:
feat(map-service): add global files including:

- configs
- exceptions
- dtos

JihwanB committed:
feat(map-service): add controllers, dtos

- DirectionsController.java
- GeocodingController.java
- DirectionsResponseDto.java
- GeocodingResponseDto.java

JihwanB committed:
feat(map-service): add services

- DirectionsService.java
- DirectionsServiceImpl.java
- GeocodingService.java
- GeocodingServiceImpl.java

JihwanB committed:
feat(map-service): add DTOs for external api response

- DirectionsApiResponse.java
- GeocodingApiResponse.java

JihwanB committed:
feat(map-service): add exceptions

- MapException.java
- MapExceptionCode.java

JihwanB committed:
chore(map-service): adjust log level

- 예외 발생시 log의 레벨을 warn으로 변경

JihwanB committed:
feat(map-service): add application-dev.yml

- 환경변수를 사용해서 외부 api에 대한 설정 추가

JihwanB committed:
refactor(map-service): 경로 정보 API의 응답 내용 변경

- 기존에 있던 path를 제거
- tollFare 필드를 추가

Bulgogi-Pizza reviewed (approved):
엔티티가 아주 복잡하네요 고생하셨습니다.
API 문서 요구사항이 잘 지켜진 것으로 보입니다.

eureka-server와 함께 PR이 올라왔는데, 원래는 분리가 되어야 하지 않을까 싶긴 하지만, 먼저 올라온 eureka-server PR과 충돌이 없다면 이번엔 괜찮지 않을까 싶습니다.

고생하셨습니다!

hyper-log commented on map-service/src/main/java/org/onlogistics/mapservice/presentation/controller/DirectionsController.java:
```diff
@@ -0,0 +1,28 @@
+        @RequestParam("start") String start,
+        @RequestParam("end") String end,
+        @RequestParam(value = "option", required = false, defaultValue = "traoptimal") String option,
+        @RequestParam(value = "cartype", required = false, defaultValue = "2") int cartype) {
```
해당 부분이 API 명세에는 지워져 있어서 명시적으로 표시만 된다면 좋을 것 같습니다!

hyper-log replied:
주말에 너무 고생 많으셨습니다 ㅠ.ㅠ

ss0ming commented on map-service/src/main/java/org/onlogistics/mapservice/presentation/controller/DirectionsController.java:
```diff
@@ -0,0 +1,28 @@
+        @RequestParam("end") String end,
+        @RequestParam(value = "option", required = false, defaultValue = "traoptimal") String option,
+        @RequestParam(value = "cartype", required = false, defaultValue = "2") int cartype) {
+        DirectionsResponseDto resDTO = directionsService.getRoute(start, end, option, cartype);
```
사소하지만 혹시 이 변수명만 resDTO로 쓰신 이유가 있을까요?

JihwanB replied:
팀 노션의 컨벤션 페이지에서 "공통 응답 및 전역 예외 처리 컨벤션"에 적혀있는 내용을 참조했었는데, 제안해주실 다른 방안이 있다면 알려주시면 감사하겠습니다!

ss0ming replied:
제가 참고한 팀 노션 컨벤션 페이지입니다. 다시 보니 두 페이지가 다르네요! 한 번 여쭤봐야할거 같아요.
https://www.notion.so/teamsparta/DTO-1b32dc3ef51480ba8065d1d34f7f43f2

IamJuho replied:
팀 컨벤션 **확실하게** 수정했습니다! 확인 부탁드릴게요~~!

JihwanB replied:
수정 했습니다!

IamJuho commented on map-service/src/main/resources/application.yml:
이 정도 데이터는 깃 레포에 올려서 공개해도 괜찮은 걸까요?

JihwanB replied:
혹시 문제될 만한 부분이 있을까요? 저희가 아직 config-repo나 config 서버가 없어서, api 키 등은 우선 환경변수로 설정해서 올렸습니다.

JihwanB replied:
jasypt로 인코딩 진행했습니다.

ss0ming commented on map-service/src/main/java/org/onlogistics/mapservice/presentation/controller/GeocodingController.java:
```diff
@@ -0,0 +1,25 @@
+    private final GeocodingService geocodingService;
+
+    @GetMapping("/api/v1/maps/geocode")
+    public ResponseEntity<CommonResponse<?>> getCoordinates(
```
해당 코드에서 ResponseEntity<CommonResponse<?>> getCoordinates 부분을 보았을 때 저희 컨벤션 문서에서 제네릭 타입을 사용하지 않기로 논의한 것으로 기억하고 있는데 확인해 주실 수 있을까요?
https://teamsparta.notion.site/DTO-1b32dc3ef51480ba8065d1d34f7f43f2

JihwanB replied:
"공통 응답 및 전역 예외 처리 컨벤션" 여기 부분이랑 헷갈렸던 것 같습니다. 아까 그룹명 바꾼 이후로 지금 로컬에서 작업이 일시적으로 불가해서, 추후에 다시 변경해도 괜찮을까요?

IamJuho replied:
이 부분도 공통 응답 컨벤션에 수정했습니다!

JihwanB replied:
수정 했습니다!

IamJuho commented on map-service/src/main/java/org/onlogistics/mapservice/presentation/controller/DirectionsController.java:
```diff
@@ -0,0 +1,28 @@
+        @RequestParam("start") String start,
+        @RequestParam("end") String end,
+        @RequestParam(value = "option", required = false, defaultValue = "traoptimal") String option,
+        @RequestParam(value = "cartype", required = false, defaultValue = "2") int cartype) {
```
int cartype 후에 들여쓰기를 통해 매개변수의 가독성을 늘려줄 수 있을 것 같아요!!

JihwanB replied:
![image](https://github.com/user-attachments/assets/3e1d9747-74eb-4eb6-bdca-b9f31f685eb6)
이런시식으로 말씀하시는 걸까요?

IamJuho replied:
```java
    @GetMapping("/api/v1/maps/route")
    public ResponseEntity<CommonResponse<?>> getRoute(
        @RequestParam("start") String start,
        @RequestParam("end") String end,
        @RequestParam(value = "option", required = false, defaultValue = "traoptimal") String option,
        @RequestParam(value = "cartype", required = false, defaultValue = "2") int cartype
      ) {
        // ...
      }
```

이렇게 생각했습니다!

JihwanB replied:
네 감사합니다. 수정 하겠습니다!

IamJuho commented on map-service/src/main/java/org/onlogistics/mapservice/presentation/controller/GeocodingController.java:
```diff
@@ -0,0 +1,25 @@
+
+    @GetMapping("/api/v1/maps/geocode")
+    public ResponseEntity<CommonResponse<?>> getCoordinates(
+        @RequestParam("query") String query) {
```
String query 후에 들여쓰기를 통해 매개변수의 가독성을 늘려줄 수 있을 것 같아요!!

JihwanB replied:
![image](https://github.com/user-attachments/assets/32a3c8ff-1274-45c5-ae15-0beb182b8d8d)
이런식으로 말씀하시는 걸까요?

IamJuho replied:
```java
    @GetMapping("/api/v1/maps/geocode")
    public ResponseEntity<CommonResponse<?>> getCoordinates(
            @RequestParam("query") String query
    ) {
      // ...
    }
    
```

이렇게 생각했습니다!

JihwanB replied:
감사합니다. 수정하겠습니다!

IamJuho commented on map-service/src/main/java/org/onlogistics/mapservice/application/service/DirectionsServiceImpl.java:
WebClient 객체를 사용하는 역할을 서비스와 분리해준다면 확장성이 높아질 수 있을 것 같아요!
이후 Kafka로 마이그레이션 하는 과정이 더 간편해지지 않을까 생각이 드는데 어떻게 생각하시나요?!

JihwanB replied:
외부 API 호출용이라 다른 서비스에서는 반영하겠습니다!

IamJuho commented on map-service/src/main/java/org/onlogistics/mapservice/application/service/DirectionsServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+            .summary(routeSummary)
+            .path(path)
+            .build();
+    }
```
이 기능은 DirectionsResponseDto 내부에서 이뤄진다면, DirectionResponseDto의 생성에 대한 역할을 객체 본인에게 주어 유지보수성이 높아질 수 있을 것 같다는 생각이 들어요! 어떻게 생각하시나요?

JihwanB replied:
반영했습니다!

JihwanB committed:
chore(map-service): DTO 컨벤션 적용

- generics를 사용하지 않도록 수정
- 반환할 데이터명을 responseDto로 변경

IamJuho commented on map-service/src/main/java/org/onlogistics/mapservice/application/service/GeocodingServiceImpl.java:
여기서도 WebClient 객체를 사용하는 역할을 분리해줄 수 있을 것 같아요!
그렇게 된다면 중복도 줄어들 수 있어 보여요!

JihwanB replied:
마찬가지로 외부 API 호출용이라 다른 서비스에서는 반영하겠습니다!

IamJuho commented on map-service/src/main/java/org/onlogistics/mapservice/application/service/GeocodingServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+            log.error("외부 API 호출 중 오류 발생. query: {}", query, e);
+            throw e;
+        } catch (Exception e) {
+            log.error("좌표 변환 중 오류 발생. query: {}", query, e);
```
로그 값은 전역 예외 핸들러에서 처리해줄 수 있지 않을까 싶습니다!
그렇게 된다면 발생되는 예외에 대한 로그 처리는 한 곳에 모일 수 있어 유지보수성이 높아질 것 같은데 어떻게 생각하시나요?

(다른 곳에서도 예외가 발생할 때 로그를 작성해주는 코드를 핸들러로 옮길 수 있을 것 같습니다!)

JihwanB replied:
좋은 것 같습니다! 해당 내용 반영해서 수정했습니다.

IamJuho reviewed (commented):
수고하셨어요!! 👍

JihwanB committed:
chore(map-service): group명 변경

- 그룹명을 on.logistics로 변경

JihwanB committed:
style: Parameter 이후 들여쓰기 추가

JihwanB committed:
feat(map-service): add jasypt

- JasyptConfig 추가
- secret을 기존의 환경변수에서 jasypt를 활용하도록 변경

JihwanB committed:
feat(map-service): add static method to dto

- dto 생성에 대한 역할을 객체 본인에게 주도록 정적 메서드 추가

JihwanB committed:
refactor(map-service): remove warn log

- 전역 예외 핸들러에서 로그를 찍어주기에 중복으로 찍히지 않게 제거
- 외부 API 호출의 성공여부를 알 수 있게 info
레벨의 로그 추가

IamJuho reviewed (approved):
LGTM 👍

ss0ming reviewed (approved):
고생하셨습니다!
